dev-utils/update-tav-versions.js

```diff
@@ -0,0 +1,224 @@
+// the script will update the versions property according to the `update-versions`
+// configuration:
+// - range: only versions satisfiying this range will be in the update
+// - mode: how we want to pick the versions within the range. These are
+//   - latest-minors: picks 1st version of the range and all the latest minor versions
+//   - latest-majors: picks 1st version of the range and all the latest major versions
+//   - max-(n): picks 1st version, the last version and (n) versions in between them
+//
+// Usage:
+//      node dev-utils/update-tav-versions.js
+
+const { execSync } = require('child_process');
+const { readFileSync, writeFileSync } = require('fs');
+const path = require('path');
+
+const semver = require('semver');
+const yaml = require('js-yaml');
+
+const TOP = path.resolve(path.join(__dirname, '..'));
+const TAV_PATH = path.join(TOP, '.tav.yml');
+const UPDATE_PROP = 'update-versions';
+
+async function main() {
+  /** @type {Map<string, string[]>} */
+  const pkgVersMap = new Map(); // versions per package name
+  /** @type {Map<string, string[]>} */
+  const tavVersMap = new Map(); // versions per TAV configuration
+
+  const tavContent = readFileSync(TAV_PATH, { encoding: 'utf-8' });
+  /** @type {Record<string, TavConfig>} */
+  const tavConfig = yaml.load(tavContent);
+  const tavEntries = Object.entries(tavConfig);
+
+  for (const entry of tavEntries) {
+    const [name, cfg] = entry;
+
+    if (!cfg[UPDATE_PROP]) continue;
+
+    const { mode, include, exclude } = cfg[UPDATE_PROP];
+    const pkgName = cfg.name || name;
+    let pkgVersions = pkgVersMap.get(pkgName);
+
+    if (!pkgVersions) {
+      console.log(`Fetching versions for ${pkgName}`);
+      pkgVersions = JSON.parse(
+        execSync(`npm view ${pkgName} versions -j`, { encoding: 'utf-8' }),
+      );
+      pkgVersMap.set(pkgName, pkgVersions);
+    }
+
+    let versions;
+    const filteredVers = pkgVersions
+      .filter((v) => semver.satisfies(v, include))
+      .filter((v) => !exclude || !semver.satisfies(v, exclude))
+      .map(semver.parse);
+
+    console.log(
+      `Calculating ${mode} for ${name} including ${include} and excluding ${
+        exclude || 'none'
+      }`,
+    );
+    if (mode === 'latest-minors') {
+      versions = getLatestMinors(filteredVers);
+    } else if (mode === 'latest-majors') {
+      versions = getLatestMajors(filteredVers);
+    } else if (mode.startsWith('max-')) {
+      const num = Number(mode.split('-')[1]);
+      if (isNaN(num)) {
+        console.error(
+          `Error: Version selection max-n mode for TAV config ${name} invalid (${mode})`,
+        );
+        continue;
+      } else {
+        versions = getMax(filteredVers, Number(num));
+      }
+    } else {
+      console.error(
+        `Error: Version selection mode for TAV config ${name} unknown (${mode})`,
+      );
+      continue;
+    }
+
+    // Assuming `include` is always in the form ">={Lower_limit} <{Higher_Limit}"
```

Comment:
nit: Might be nice to check that and throw if it isn't the case.

Reply:
I've added the format validations at the top of the loop so the script throws before doing an expensing operation like fetching the list of versions from npm.